libtenzir/builtins/formats/feather.cpp

```diff
@@ -285,22 +287,108 @@ class callback_listener : public arrow::ipc::Listener {
+      co_yield {};
+      continue;
+    }
+    truncated_bytes += payload->size();
+    if (payload->size() < required_size) {
+      if (truncated_bytes != 0 and payload->size() != 0) {
+        // Ideally this always would be just a warning, but the stream decoder
+        // happily continues to consume invalid bytes. E.g., trying to read a
+        // JSON file with this parser will just swallow all bytes, emitting this
+        // one error at the very end. Not a single time does consuming a buffer
+        // actually fail. We should probably look into limiting the memory usage
+        // here, as the stream decoder will keep consumed-but-not-yet-converted
+        // buffers in memory.
+        diagnostic::warning("truncated {} trailing bytes", truncated_bytes)
+          .severity(decoded_once ? severity::warning : severity::error)
+          .emit(ctrl.diagnostics());
+      }
+      co_return;
+    }
+    auto arrow_buffer = as_arrow_buffer(std::move(payload));
+    buffered_schema_data.push_back(arrow_buffer);
+    auto decode_result = schema_stream_decoder.Consume(std::move(arrow_buffer));
+    if (!decode_result.ok()) {
+      diagnostic::error("{}", decode_result.ToStringWithoutContextLines())
+        .note("failed to decode the byte stream into a record batch")
+        .emit(ctrl.diagnostics());
+      co_return;
+    }
+    if (schema_listener->decoded_schema) {
+      decoded_once = false;
+      schema = type::from_arrow(*schema_listener->decoded_schema);
+      break;
+    }
+  }
+  auto indices = std::vector<tenzir::offset>{};
+  if (!selection.inner.fields_of_interest.empty()) {
+    for (const auto& field : selection.inner.fields_of_interest) {
+      for (auto index : schema.resolve(field)) {
+        if (index.size() > 1) {
+          further_selection_needed = true;
+        }
+        read_options.included_fields.push_back(static_cast<int>(index[0]));
+        indices.push_back(std::move(index));
+      }
+    }
+    std::sort(indices.begin(), indices.end());
+    indices.erase(std::unique(indices.begin(), indices.end()), indices.end());
+    std::sort(read_options.included_fields.begin(),
+              read_options.included_fields.end());
+    read_options.included_fields.erase(
+      std::unique(read_options.included_fields.begin(),
+                  read_options.included_fields.end()),
+      read_options.included_fields.end());
+    for (size_t i = 0; i < indices.size(); i++) {
+      indices[i][0] = i;
+    }
+  }
+  auto listener = std::make_shared<callback_listener>();
+  auto stream_decoder = arrow::ipc::StreamDecoder(listener, read_options);
+  for (auto&& buffer : buffered_schema_data) {
+    auto decode_result = stream_decoder.Consume(buffer);
+    if (!decode_result.ok()) {
+      diagnostic::error("{}", decode_result.ToStringWithoutContextLines())
+        .note("failed to decode the byte stream into a record batch")
+        .emit(ctrl.diagnostics());
+      co_return;
+    }
+  }
```

Comment:
This code is almost exactly duplicated from right below. Can we make this easier to maintain?

Reply:
I tried for a few minutes, but I couldn't come up with something better. So maybe just leave it as-is, as annoying as that is.